### PR TITLE
xplat intro 2 mvc - add a field

### PR DESCRIPTION
**Pull request by Rick-Anderson**
[Review URL](https://review.docs.microsoft.com/en-us/aspnet/core/tutorials/first-mvc-app-xplat/new-field?branch=pr-en-us-3197) - compare to [migrations version](https://docs.microsoft.com/en-us/aspnet/core/tutorials/first-mvc-app/new-field)

@bricelam Can you suggest a better workflow? Deleting source files (the migrations files) and a database when you make schema changes sees klugy.

### CODE REVIEW

**Comment by bricelam** [comment]
I would make sure the following is called during app start.
```C#
db.Database.EnsureCreated();
```
This will create the database if it doesn't exist.

Then whenever you change your model, you can just run this command.

    dotnet ef database drop

Note, we've talked about making a command for `EnsureCreated`, but we need to sort out some usability concerns. (See aspnet/EntityFramework#7840)

**Comment by bricelam** [comment]
And in case it's not obvious: Don't use Migrations.

File: aspnetcore/tutorials/first-mvc-app-xplat/new-field.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+This tutorial will add a new field to the `Movies` table. One option is to use [EF Code First Migrations](http://docs.efproject.net/en/latest/platforms/aspnetcore/new-db.html) to add a new field to the model and migrate that change to the database - but SQLlite does not support many migration schema operations, so only very simply migrations are possible. See [SQLite Limitations](https://docs.microsoft.com/ef/core/providers/sqlite/limitations) for more information.
+
+For this tutorial we'll drop our database and create a new one when we change the schema. This workflow works well early in development when we don't have any production data to perserve.
+
+See [Migrations limitations workaround](https://docs.microsoft.com/ef/core/providers/sqlite/limitations?branch=Rick-Anderson-patch-3#migrations-limitations-workaround) for one approch to using migrations with SQLite.
+
+## Adding a Rating Property to the Movie Model
```
Comment (tdykstra) [inline aspnetcore/tutorials/first-mvc-app-xplat/new-field.md]: The doc now has only a single h2.  You could add an h2 with a title like "Add the new property to the database"  right before the discussion of choices for updating the database.

The Seed data update is out of place in that case, but I'd move it to right after the model update anyway.

File: aspnetcore/tutorials/first-mvc-app-xplat/new-field.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+The app won't work until we update the DB to include the new field. If you run it now, you'll get the following `SqliteException`:
+
+`SqliteException: SQLite Error 1: 'no such column: m.Rating'.``
+
+You're seeing this error because the updated Movie model class is different than the schema of the Movie table of the existing database. (There's no `Rating` column in the database table.)
+
+There are a few approaches to resolving the error:
+
+1. Have the Entity Framework automatically drop and re-create the database based on the new model class schema. This approach is very convenient early in the development cycle when you are doing active development on a test database; it allows you to quickly evolve the model and database schema together. The downside, though, is that you lose existing data in the database — so you don't want to use this approach on a production database! Using an initializer to automatically seed a database with test data is often a productive way to develop an app.
```
Comment (tdykstra) [inline aspnetcore/tutorials/first-mvc-app-xplat/new-field.md]: I don't think it automatically both drops and re-creates

File: aspnetcore/tutorials/first-mvc-app-xplat/new-field.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+ms.prod: asp.net-core
+uid: tutorials/first-mvc-app-xplat/new-field
+---
+# Adding a New Field
+
+By [Rick Anderson](https://twitter.com/RickAndMSFT)
+
+When you use EF Code First to automatically create a database, Code First adds a table to the database to help track whether the schema of the database is in sync with the model classes it was generated from. If they aren't in sync, EF throws an exception. This makes it easier to find inconsistent database/code issues.
+
+This tutorial will add a new field to the `Movies` table. One option is to use [EF Code First Migrations](http://docs.efproject.net/en/latest/platforms/aspnetcore/new-db.html) to add a new field to the model and migrate that change to the database - but SQLlite does not support many migration schema operations, so only very simply migrations are possible. See [SQLite Limitations](https://docs.microsoft.com/ef/core/providers/sqlite/limitations) for more information.
```
Comment (tdykstra) [inline aspnetcore/tutorials/first-mvc-app-xplat/new-field.md]: Consider presenting here what you're going to do, i.e., drop and re-create the database, then in a separate paragraph, discuss migrations, ending with a link to that limitations/workarounds doc.  As it stands, it's a little confusing hearing about migrations, then we're not going to use it, then we're presented with options for migrations workarounds.

An alternative is to put the migrations discussion below after #3 in the "There are a few approaches to resolving the error."  Would be a good place to both introduce migrations and explain why you're not using it.

**Comment by Rick-Anderson** [comment]
@tdykstra  ready for review again.

**Comment by tdykstra** [comment]
@Rick-Anderson 

> ready for review again.

It doesn't look like you've pushed any new commits (?)

**Comment by Rick-Anderson** [comment]
@tdykstra ready for review again - really.

File: aspnetcore/tutorials/first-mvc-app-xplat/new-field.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+Once your app is deployed and you have data that you need to perserve, you can't drop your DB when you need to change the schema. Entity Framework [Code First Migrations](http://docs.efproject.net/en/latest/platforms/aspnetcore/new-db.html) allows you to update your schema and migrate the database without losing data. Migrations is a popular feature when using SQL Server, but SQLlite does not support many migration schema operations, so only very simply migrations are possible. See [SQLite Limitations](https://docs.microsoft.com/ef/core/providers/sqlite/limitations) for more information.
+
+## Adding a Rating Property to the Movie Model
+
+Open the *Models/Movie.cs* file and add a `Rating` property:
+
+[!code-csharp[Main](../first-mvc-app/start-mvc/sample/MvcMovie/Models/MovieDateRating.cs?highlight=11&range=7-18)]
+
+Because you've added a new field to the `Movie` class, you also need to update the binding white list so this new property will be included. In *MoviesController.cs*, update the `[Bind]` attribute for both the `Create` and `Edit` action methods to include the `Rating` property:
```
Comment (tdykstra) [inline aspnetcore/tutorials/first-mvc-app-xplat/new-field.md]: white list --> whitelist

File: aspnetcore/tutorials/first-mvc-app-xplat/new-field.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+```csharp
+[Bind("ID,Title,ReleaseDate,Genre,Price,Rating")]
+   ```
+
+You also need to update the view templates in order to display, create and edit the new `Rating` property in the browser view.
```
Comment (tdykstra) [inline aspnetcore/tutorials/first-mvc-app-xplat/new-field.md]: comma after create

File: aspnetcore/tutorials/first-mvc-app-xplat/new-field.md [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+Edit the */Views/Movies/Index.cshtml* file and add a `Rating` field:
+
+[!code-HTML[Main](../first-mvc-app/start-mvc/sample/MvcMovie/Views/Movies/IndexGenreRating.cshtml?highlight=17,39&range=24-64)]
+
+Update the */Views/Movies/Create.cshtml* with a `Rating` field.
+
+The app won't work until we update the DB to include the new field. If you run it now, you'll get the following `SqliteException`:
+
+`SqliteException: SQLite Error 1: 'no such column: m.Rating'.``
```
Comment (tdykstra) [inline aspnetcore/tutorials/first-mvc-app-xplat/new-field.md]: formatting isn't working on this, suggest you make it a fenced code block (inside triple-backticks) or blockquote (line starts with open angle bracket)

File: aspnetcore/tutorials/first-mvc-app-xplat/new-field.md [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+You're seeing this error because the updated Movie model class is different than the schema of the Movie table of the existing database. (There's no `Rating` column in the database table.)
+
+There are a few approaches to resolving the error:
+
+1. Drop the database and have the Entity Framework automatically re-create the database based on the new model class schema. This approach is very convenient early in the development cycle when you are doing active development on a test database; it allows you to quickly evolve the model and database schema together. The downside, though, is that you lose existing data in the database — so you don't want to use this approach on a production database! Using an initializer to automatically seed a database with test data is often a productive way to develop an app.
```
Comment (tdykstra) [inline aspnetcore/tutorials/first-mvc-app-xplat/new-field.md]: The explanation here ("This approach is very convenient ..." ) is the same as what you have in the intro at top of page; I don't think you need it in both places.

File: aspnetcore/tutorials/first-mvc-app-xplat/new-field.md [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+2. Explicitly modify the schema of the existing database so that it matches the model classes. The advantage of this approach is that you keep your data. You can make this change either manually or by creating a database change script.
```
Comment (tdykstra) [inline aspnetcore/tutorials/first-mvc-app-xplat/new-field.md]: "Manually" might be better than "Explicitly" here. In any case, in the last sentence it's not clear what the difference is between "manually" and "create a script"; if I create a script I'm doing it manually, it isn't automatic.

File: aspnetcore/tutorials/first-mvc-app-xplat/new-field.md [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+For this tutorial, we'll drop and re-create the database when the schema changes. Run the following command from a terminal to drop the db:
+
+`dotnet ef database drop`
+
+Update the `SeedData` class so that it provides a value for the new column. A sample change is shown below, but you'll want to make this change for each `new Movie`.
+
+[!code-csharp[Main](../first-mvc-app/start-mvc/sample/MvcMovie/Models/SeedDataRating.cs?name=snippet1&highlight=6)]
+
+Run the app and verify you can create/edit/display movies with a `Rating` field. You should also add the `Rating` field to the `Edit`, `Details`, and `Delete` view templates.
```
Comment (tdykstra) [inline aspnetcore/tutorials/first-mvc-app-xplat/new-field.md]: These sentences are out of order - they have you editing and displaying before they have you updating the update the edit an details views.

**Comment by bricelam** [comment]
👍 Much less klugy

**Pull request merged**